pkg/networkservice/common/heal/client.go

```diff
@@ -19,41 +19,208 @@ package heal
-
 type healClient struct {
-	ctx context.Context
-	cc  networkservice.MonitorConnectionClient
+	ctx              context.Context
+	cc               networkservice.MonitorConnectionClient
+	initOnce         sync.Once
+	initErr          error
+	connCache        map[string]*networkservice.Connection
+	conCacheExecutor serialize.Executor
+	expectedConns    verificationRequestMap
 }
 
 // NewClient - creates a new networkservice.NetworkServiceClient chain element that inform healServer about new client connection
 func NewClient(ctx context.Context, cc networkservice.MonitorConnectionClient) networkservice.NetworkServiceClient {
 	return &healClient{
-		ctx: ctx,
-		cc:  cc,
+		ctx:       ctx,
+		cc:        cc,
+		connCache: make(map[string]*networkservice.Connection),
 	}
 }
 
 func (u *healClient) Request(ctx context.Context, request *networkservice.NetworkServiceRequest, opts ...grpc.CallOption) (*networkservice.Connection, error) {
+	u.initOnce.Do(func() {
+		errCh := make(chan error, 1)
+		pushFunc := healRequestFunc(ctx)
+		go u.listenToConnectionChanges(pushFunc, request.GetConnection().GetCurrentPathSegment(), errCh)
+		u.initErr = <-errCh
+	})
+	// if initialization failed, then we want for all subsequent calls to Request() on this object to also fail
+	if u.initErr != nil {
+		return nil, u.initErr
+	}
+
+	verificationRequest := u.addToExpectedConnections(request.GetConnection())
+	if verificationRequest != nil {
+		defer close(verificationRequest.stopCh)
+	}
+
 	conn, err := next.Client(ctx).Request(ctx, request, opts...)
+	if err != nil {
+		return nil, err
+	}
 
-	registerClient := registerClientFunc(ctx)
-	if err == nil && registerClient != nil {
-		registerClient(u.ctx, conn, u.cc)
+	if verificationRequest != nil {
+		err = u.confirmConnectionSuccess(conn, verificationRequest)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		u.addConnectionToMonitor(conn)
 	}
-	return conn, err
+
+	return conn, nil
 }
 
 func (u *healClient) Close(ctx context.Context, conn *networkservice.Connection, opts ...grpc.CallOption) (*empty.Empty, error) {
+	if cachedConn, ok := u.removeConnectionFromMonitor(conn); ok {
+		conn = cachedConn
+	} else {
+		log.FromContext(ctx).WithField("healClient", "Close").Warnf("can't find connection in cache, id=%v", conn.GetId())
+	}
+
 	return next.Client(ctx).Close(ctx, conn, opts...)
 }
+
+func (u *healClient) listenToConnectionChanges(heal healRequestFuncType, currentPathSegment *networkservice.PathSegment, errCh chan error) {
+	monitorClient, err := u.cc.MonitorConnections(u.ctx, &networkservice.MonitorScopeSelector{
+		PathSegments: []*networkservice.PathSegment{{Name: currentPathSegment.Name}, {Name: ""}},
+	})
+	if err != nil {
+		errCh <- errors.Wrap(err, "MonitorConnections failed")
+		return
+	}
+	close(errCh)
+
+	for {
+		event, err := monitorClient.Recv()
+		if err != nil {
+			if heal != nil {
+				<-u.conCacheExecutor.AsyncExec(func() {
+					for id := range u.connCache {
+						heal(id, true)
```

Comment:
We probably want to pass updated path from `connCache` to the `heal` to make it work properly. I suppose that it is not enough to pass only `id` because in such case `heal` will be invoked with a new client and there will be no `connCache`.